soda/scientific/soda/scientific/anomaly_detection/anomaly_detector.py

```diff
@@ -0,0 +1,263 @@
+        "anomalyPredictedValue": 13,
+        "anomalyErrorSeverity": "warn",
+        "anomalyErrorCode": "inferred_daily_frequency",
+        "feedback": {},
+    }
+]
+
+
+class AnomalyDetector:
+    def __init__(self, measurements, check_results, logs: Logs):
+        self._logs = logs
+        self.df_measurements = self._parse_historical_measurements(measurements)
+        self.df_check_results = self._parse_historical_check_results(check_results)
+        self.params = self._parse_params()
+
+    @staticmethod
+    def _parse_historical_measurements(measurements: Dict[str, List[Dict[str, Any]]]) -> pd.DataFrame:
+        if measurements:
+            parsed_measurements = AnomalyHistoricalMeasurements.parse_obj(measurements)
+            _df_measurements = pd.DataFrame.from_dict(parsed_measurements.dict()["results"])
+            return _df_measurements
+        else:
+            raise ValueError("No historical measurements found.")
+
+    def _parse_historical_check_results(self, check_results: Dict[str, List[Dict[str, Any]]]) -> pd.DataFrame:
+        if check_results.get("results"):
+            parsed_check_results = AnomalyHistoricalCheckResults.parse_obj(check_results)
+            _df_check_results = pd.DataFrame.from_dict(parsed_check_results.dict()["results"])
+            return _df_check_results
+        else:
+            self._logs.info(
+                "No past check results found. This could be because there are no past runs of "
+                "Anomaly Detection for this check yet."
+            )
+            parsed_check_results = AnomalyHistoricalCheckResults(results=[AnomalyResult()])
+            _df_measurements = pd.DataFrame.from_dict(parsed_check_results.dict()["results"])
+            return _df_measurements
+
+    def _convert_to_well_shaped_df(self) -> pd.DataFrame:
+        if not self.df_check_results.empty:
+            self._logs.debug("Got test results from data request. Merging it with the measurements")
+            df = self.df_measurements.merge(
+                self.df_check_results,
+                how="left",
+                left_on="id",
+                right_on="measurementId",
+                suffixes=("", "_tr"),
+            )
+        else:
+            df = self.df_measurements.copy()
+
+        # Flatten diagnostics dictionary
+        if "diagnostics" in df.columns:
+            df_flattened = self.flatten_df(df.copy(), "diagnostics")
+
+        column_maps = self.params["request_params"]["columns_mapping"]
+        df_flattened = df_flattened[df_flattened.columns[df_flattened.columns.isin(list(column_maps.keys()))]]
+        df_flattened = df_flattened.rename(columns=column_maps)  # type: ignore
+        df_flattened["ds"] = pd.to_datetime(df_flattened["ds"])  # type: ignore
+        df_flattened["ds"] = df_flattened["ds"].dt.tz_localize(None)
+        return df_flattened
+
+    @staticmethod
+    def flatten_df(df: pd.DataFrame, target_col_name: str) -> pd.DataFrame:
```

Comment:
as far as I remember, this method is just used in this class. So, to make it consistent we can change the name from flatten_df to _flatten_df